docs/architecture/adr-versions.md

```diff
@@ -0,0 +1,219 @@
+# Protocol and Chain Versions
+
+The Software Version is covered by SemVer and described elsewhere.
+It is not relevant to the protocol description, suffice to say that if any protocol version
+changes, the software version changes. Software version is included in NodeInfo for convenience/diagnostics.
+
+Here we focus on protocol and chain versions.
+
+## Requirements
+
+We need to version blockchains across protocols, networks, forks, etc.
+We need to do it in a way that is scalable and maintainable - we can't just litter
+the code with conditionals.
+We need chain identifiers and descriptions so we can talk about a multitude of chains in a meaningful way.
+
+Let's start with versions.
+
+### Protocol Version
+
+The complete version of the protocol contains the following sub-versions:
+BlockVersion, P2PVersion, AppVersion. These versions reflect the major sub-components
+of the software that are likely to evolve together, at different rates, and in different ways,
+as described below.
+
+#### BlockVersion
+
+- All tendermint hashed data-structures (headers, votes, txs, responses, etc.).
+	- Note the semantic meaning of a transaction may change according to the AppVersion,
+		but the way txs are merklized into the header is part of the BlockVersion
+- It should be the least frequent/likely to change.
+	- Tendermint should be stabilizing - it's just Atomic Broadcast.
+	- We can start considering for Tendermint v2.0 in a year
+- For now, assume that BlockVersion will not change in the life of a given chain, except by critical emergency.
+- To effectively change it, dump state and start a new blockchain with the new BlockVersion.
+- It's easy to determine the version of a block from its serialized form
+
+#### P2PVersion
+
+- All peer messaging (messages, detectable behaviour)
+- Will change gradually as reactors evolve to improve performance and support new features
+	- eg proposed new message types BatchTx in the mempool and HasBlockPart in the consensus
+- It's easy to determine the version of a peer from its first serialized message/s
+- New versions must be compatible with at least one old version to allow gradual upgrades
+
+#### AppVersion
+
+- The ABCI state machine (txs, begin/endblock behaviour, commit hashing)
```

Comment:
Why we care about this at Tendermint level? Isn't Tendermint app agnostic? Even for state syncing, we need to have a mechanisms that transfer bytes around but we don't care what is inside. It is app business.

Reply:
I think we want to reject peers with a different app version